Added Proposal Kafka Preview Feature Support in Strimzi (#167)
strimzi/proposals: 108-kafka-preview-features.md

## Declarative Management of Kafka Preview Features in Strimzi

## Current Situation:

> Kafka preview features (such as kafka.shared.consumer) need to be manually enabled using the kafka-features.sh or the Kafka Admin API.

Reviewer: ```suggestion
Kafka preview features (such as `kafka.shared.consumer`) need to be manually enabled using the ¨kafka-features.sh` or the Kafka Admin API.
```

This works in static clusters but presents issues in dynamic, testing or GitOps-managed environments.
There is currently no way to express these features declaratively within the Kafka custom resource.
Without declarative support, the following issues arise:

> 
> - Feature drift from manual changes.
> 
> - Loss of settings in ephemeral environments.
> 
> - Inability to track changes through Git.

Reviewer: ```suggestion
* Feature drift from manual changes.
* Loss of settings in ephemeral environments.
* Inability to track changes through Git.
```

Kafka Preview Features Explained
Kafka preview features are opt-in capabilities introduced in newer Kafka versions to allow early access to experimental or upcoming functionality.

> These features are disabled by default and must be explicitly enabled on each broker to become usable.

Reviewer: So, what happens when the features GA? How will we validate them? How will the error handling look like?

Author: Validation:

1. During reconciliation, Strimzi will query the broker’s supported features via the Kafka Admin API (DescribeFeatures).
2. If the feature is GA and still listed in .spec.kafka.previewFeatures, the operator will check whether:
3. Kafka still allows managing it through the UpdateFeatures API.
4. The requested version is valid for the current Kafka version.

Error Handling:
- If the feature is GA, **enabled by default**, and cannot be disabled, the operator will skip applying the change.Record a CannotBeDisabled condition in .status with a reason explaining that the feature is GA and mandatory.

- If the feature is **GA but supports version changes or disabling**, the operator will reconcile as per the CR.
- If the CR requests an unsupported version (either new or not available in this version), the operator will--Skips applying the change.
- Record a VersionNotSupported condition in .status listing the available versions from Kafka’s feature metadata.

Reviewer: TBH, I'm not sure it is that simple.

I can see how when the feature moves to GA we just issue some warning condition. But if it enabled some feature and that feature version is not supported because it was replaced with a new version, it sounds like ignoring it by throwing a warning would essentially just disable the feature? That seems wrong if there is newer version? What does Kafka itself do in such situation?

Also, I think this is where you need to introduce proper context from how the Kafka community expects to use these _features_. Will each Kafka version support multiple versions of some feature (it sounds like it from the API you propose and that Kafka uses?)? Is the _GA_ just a special version of the feature?

I think understanding it properly is important. And if it is not clear how Kafka will use it, we might not be able to decide on this without screwing up our API. So we would be blocked. To be honest, I think I didi not properly understood how this will evolve (maybe I missed somehting, maybe the problem is that Kafka community itself did not understood how this will evolve) when I wrote https://github.com/strimzi/proposals/blob/main/061-kraft-upgrades-and-downgrades.md as the metadata version seems to be related to this as well.

_As a sidenote: You seem to talk about the status conditions in response to multiple comments. But there is nothing about it in the proposal. If that is how you plan to deal with these, it is important to cover it there as that is the only way how to have the discussion centered in a single place. And not have the discussion independently in multiple places and possibly lead to conflicting outcomes._

Reviewer: > If Kafka allows multiple versions of the same feature to coexist (as the API suggests), the operator will treat GA as the “stable” version and attempt to reconcile to the closest supported version available.

That is not really how declarative management works. You cannot just decide to use something what the operator things is _similar_ and ignore the difference. There is also probably no way how to determine what is the _closest_ version.

> Strimzi will not silently disable the feature — instead, it will record a VersionNotSupported condition in .status and list which versions Kafka still supports, so the user is guided to update the CR. If Kafka makes the GA version mandatory and non-configurable, the operator will skip reconciling that field and surface a CannotBeDisabled condition in .status to make the user aware.

So, will this fail the reocnciliation and for example block an upgrade?

Kafka stores the state of these features in the metadata quorum (ZooKeeper or KRaft, depending on the deployment).
The feature state includes:

> 
> - Feature name (e.g., kafka.shared.consumer)
> 
> - Version 
> 
> - Enabled/disabled flag

Reviewer: ```suggestion
* Feature name (e.g., kafka.shared.consumer)
* Version 
*  Enabled/disabled flag
```

Currently, these features are enabled via:

> 
> The kafka-features.sh  tool
> 
> The Kafka Admin API (Feature API)

Reviewer: ```suggestion
* The `kafka-features.sh` tool
* The Kafka Admin API (Feature API)
```

However, these tools require manual execution after cluster creation, which is incompatible with automated GitOps workflows or testing environments.

This proposal introduces a way to declare desired preview features via the Kafka CR, allowing the Strimzi operator to apply and reconcile the correct state automatically.

> 
> 
> ## Motivation:

Reviewer: ```suggestion

## Motivation:

```

These features are disabled by default and must be explicitly enabled on each broker in the cluster.
In environments where clusters are frequently recreated, such as CI/CD pipelines or testing setups, manually enabling these features is both error-prone and hard to maintain.

- In Kafka, preview features are managed per-broker and stored in ZooKeeper or KRaft metadata.
- When a broker starts, it checks the feature state before making it available to Clients.
- If a feature is disabled or at a lower version, functionality depending on it may fail.

In GitOps workflows, lack of declarative configuration breaks the principle of having the full desired cluster state stored in version control.
As a result:

- Teams cannot easily maintain consistent environments across dev/staging/prod.

- Manual intervention is needed after each deployment or cluster recreation.

By supporting preview features through the Kafka CR, Strimzi can offer a seamless, GitOps-compatible, and maintainable experience for all cluster types.

Proposal:
I propose to add native declarative support for preview features in the Kafka CR.

> CRD Changes

Reviewer 2: ```suggestion
### CRD Changes
```

> Extend the Kafka CRD with a new previewFeatures section under .spec.kafka:

Reviewer 2: ```suggestion
Extend the Kafka CRD with a new preview features section under `.spec.kafka`:
```

spec:
  kafka:
    version: 3.7.0

>     previewFeatures:

Reviewer: Are these 100% previews of new features only?

Author: The .spec.kafka.previewFeatures field will only control features that Kafka exposes via the Feature API — whether they are in preview or already GA but still configurable.

1. If the feature is GA but still configurable:
The operator will allow the user to set it explicitly in previewFeatures even though the name might be misleading, we can rename it to kafkaFeatures if needed.

2. If the feature is GA and not configurable anymore:
The operator will ignore the CR entry for that feature, set a Warning condition in .status explaining that it’s GA and mandatory, and proceed with the rest.

3. If the feature is GA with multiple versions:
Users can still request a specific version in .spec.kafka.previewFeatures and the operator will attempt to set it if Kafka permits.

      - name: kafka.shared.consumer
        enabled: true

>         version: 1

Reviewer: Format this as code with ```yaml

Operator Behaviour
1) On reconciliation, Strimzi queries the current preview feature state using the Kafka Admin API.

2) If a mismatch is detected, the Operator applies the desired state using the Admin API or kafka-features.sh.

> 3) Changes are reconciled continuously to prevent drift.

Reviewer 3: Also, how it will work during upgrades of the Kafka version? Is there a way how to check what are the available features that can be enabled and in case that in the list of the `previewFeatures` inside the `.spec.kafka` will be something wrong (not available anymore or graduated to GA), should it be ignored and mentioned in the status of the Kafka CR?

Reviewer 3: The same thing should be with the downgrade, some of the features will not be available if you will do downgrade of the Kafka cluster from one version to the other, so that should be handled as well.

Author: > Also, how it will work during upgrades of the Kafka version? Is there a way how to check what are the available features that can be enabled and in case that in the list of the `previewFeatures` inside the `.spec.kafka` will be something wrong (not available anymore or graduated to GA), should it be ignored and mentioned in the status of the Kafka CR?

 Upgrade Scenario:

Queries the broker’s available features via Admin API.

Compares them with .spec.kafka.previewFeatures.

For each entry in .spec:

If still preview and supported → apply as requested.

If graduated to GA → skip enabling (Kafka enables it by default); record in .status with reason=FeatureGraduatedToGA.

If removed → skip enabling; record in .status with reason=FeatureUnavailable.

The cluster stays running without error, but users get visible feedback in status.conditions.

is what i was thinking -open to suggestions

Author: > The same thing should be with the downgrade, some of the features will not be available if you will do downgrade of the Kafka cluster from one version to the other, so that should be handled as well.

Downgrade Scenario:

If a feature is no longer supported in the downgraded version, the operator disables it via Admin API and records in .status with reason=FeatureUnsupportedInVersion.

is what i was thinking -open to suggestions

Reviewer 3: Yeah you should mention it in the proposal, so it is covered.

Benefits

> - Declarative and GitOps-friendly.
> 
> - Drift protection through automatic reconciliation.
> 
> - Consistent environments across deployments.

Reviewer 3: ```suggestion
- Declarative and GitOps-friendly.
- Drift protection through automatic reconciliation.
- Consistent environments across deployments.
```

But I guess you already mentioned the benefits before.

Considerations
- Requires changes to the Kafka CRD schema.

> - Operator logic must be extended to handle feature state reconciliation.

Reviewer: You should provide details of how exactly will the operator do it -> which phase, which classes, etc. Also, do we need some trakcing of the current state of these features?

> - Validation rules needed to ensure correct feature name, version, and enabled state.

Reviewer 3: How the validation will look like? How the user will be informed that there is something wrong? Also, should we keep some information about enabled features inside the `.status` section of the Kafka CR?

Author: The operator will validate requested features before applying them. If there's an invalid feature name, unsupported version, or other mismatch, the operator will:

Skip the problematic entry to avoid cluster failure, and

Record an entry under .status.previewFeatures including: feature name, requested versus applied version, state (Enabled, Skipped, etc.), and a clear reason (InvalidFeatureName, UnsupportedVersion).
This lets users immediately see what was applied and why something failed.

 - name: kafka.old.preview
      requestedVersion: 1
      appliedVersion: null
      state: Skipped
      reason: "Feature not available in Kafka 3.5."
    - name: kafka.new.featureX
      requestedVersion: 2
      appliedVersion: null
      state: Skipped
      reason: "Invalid feature name."
      
      
    hope this answers your question

Reviewer 3: That's good, could you please add it to the proposal? I guess it's important to know how it will be handled. Thanks :)

Limitations

- Not declarative through the Kafka CR.

- No automatic reconciliation or drift protection.

> - More error-prone due to lack of CR validation.

Author: sorry, my bad ---this limitation was for the rejected version

## Affected Projects:
- strimzi-kafka-operator — Kafka CRD schema and reconciliation logic.

> - strimzi-images — Potential updates to include Admin API handling for preview features.

Reviewer 3: I guess the "potential" is wrong word TBH.
With every change to the code (and every build of the image) you will have some updates in comparison with the current images.
Also, it's not a project, so I would remove this.

## Compatibility:

> The previewFeatures field will be optional, and clusters without it will behave as they do today.

Reviewer 3: ```suggestion
The `previewFeatures` field will be optional, and clusters without it will behave as they do today.
```

## Rejected alternatives:
External Script or Init Container
- A user-managed init container or script could run kafka-features.sh at broker startup.

> - This would be configured via a ConfigMap, environment variables, or a mounted script.

Reviewer 3: This is good, but you should maybe mention why it is rejected alternative.

Author: right
i shall add that too

## Declarative Management of Kafka Preview Features in Strimzi

## Current Situation:

Kafka preview features (such as `kafka.shared.consumer`) need to be manually enabled using the ¨kafka-features.sh` or the Kafka Admin API.

> This works in static clusters but presents issues in dynamic, testing or GitOps-managed environments.

Author: By static clusters, I mean long-lived, relatively stable Kafka environments where the configuration is applied once and rarely changes—for example, production clusters that are provisioned manually or with minimal automation. In these cases, enabling a preview feature manually through kafka-features.sh or the Admin API is manageable, since the cluster is not frequently recreated.

By dynamic clusters, I mean short-lived or continuously reconciled environments, such as clusters for testing, CI/CD pipelines, or GitOps-managed deployments where the operator ensures the declared state is always applied. In these scenarios, manually running kafka-features.sh isn’t practical because the feature enablement would be lost on cluster re-creation or drifted from the GitOps source of truth.

> This works in static clusters but presents issues in dynamic, testing or GitOps-managed environments.

Reviewer: I do not think this is a terminology we use anywhere, so you should explain it in the proposal.

However, I think you got one thing wrong:
* The _static_ production clusters often use GitOps and often cannot use the `kafka-features.sh` command because the only proper and approved way to change things is through GitOps processes (i.e. they would for example need to manage it through a Kubernetes Job calling `kafka-features.sh` which is submitted through GitOps).
* The _dynamic_ shortlived clusters would nto want to use `kafka-features.sh` every five minutes to set it. But then - you IMHO do not use that often GitOps to create a cluster for a very short time and then delete it again through GitOps. So the GitOps part of tje argument here - I do not get it.

> This works in static clusters but presents issues in dynamic, testing or GitOps-managed environments.

Author: I see your point about the terminology not being commonly used, and I’ll adjust the proposal to avoid introducing “static/dynamic” and instead describe clusters in terms of lifecycle and management style.

You’re also right that production clusters often rely on GitOps as the only approved mechanism for changes, so running kafka-features.sh directly wouldn’t fit those processes. I’ll reframe the explanation to emphasize that:
Prod clusters: usually GitOps-managed, requiring any feature enablement to be expressed declaratively, rather than direct CLI commands.

Non-prod clusters : recreated frequently, so manual commands don’t persist across rebuilds. While GitOps is less common here, automation via CI pipelines or scripts is needed instead of manual steps.

There is currently no way to express these features declaratively within the Kafka custom resource.
Without declarative support, the following issues arise:
* Feature drift from manual changes.
* Loss of settings in ephemeral environments.
* Inability to track changes through Git.

### Kafka Preview Features Explained
Kafka preview features are opt-in capabilities introduced in newer Kafka versions to allow early access to experimental or upcoming functionality.
Early access or Preview features that are not production-ready yet, are disabled by default and must be explicitly enabled on each broker to become usable.

Kafka stores the state of these features in the metadata quorum (ZooKeeper or KRaft, depending on the deployment).
The feature state includes:
* Feature name (e.g., kafka.shared.consumer)
* Version 

> *  Enabled/disabled flag

Author: I completely agree that we cannot and should not change the Kafka API itself. My intention with the “enabled/disabled” simplification is not to introduce a new abstraction that overrides Kafka’s semantics, but rather to describe how Strimzi might present the existing finalized version level in a more user-friendly way. Internally, it would still rely on the Kafka Admin API fields (FinalizedVersionLevel, SupportedMinVersion, etc.) without altering their meaning. I’ll make sure to clarify this better in the proposal so it’s clear that we’re only mapping what Kafka already exposes, not creating a divergent API surface.

- In Kafka, preview features are managed per-broker and stored in ZooKeeper or KRaft metadata.
- When a broker starts, it checks the feature state before making it available to Clients.
- If a feature is disabled or at a lower version, functionality depending on it may fail.

In GitOps workflows, lack of declarative configuration breaks the principle of having the full desired cluster state stored in version control.
As a result:

* Teams cannot easily maintain consistent environments across dev/staging/prod.
* Manual intervention is needed after each deployment or cluster recreation.

By supporting preview features through the Kafka CR, Strimzi can offer a seamless, GitOps-compatible, and maintainable experience for all cluster types.

## Proposal
I propose to add native declarative support for preview features in the Kafka CR.

CRD Changes
Extend the Kafka CRD with a new previewFeatures section under .spec.kafka:

spec:
  kafka:
    version: 3.7.0
    previewFeatures:
      - name: share.version
        enabled: true
        version: 1

### Operator Behaviour
1) On reconciliation, Strimzi queries the current preview feature state using the Kafka Admin API.

> 2) If a mismatch is detected, the Operator applies the desired state using the Admin API or `kafka-features.sh`.

Author: You’re absolutely right — the Operator should rely on the Kafka Admin API rather than invoking kafka-features.sh. The reference to kafka-features.sh in the proposal was meant only as an example of how users interact with preview features, not as an implementation detail for Strimzi. For the reconciliation logic itself, the plan is to query and update feature states through the Admin API.
